### ISSUE TITLE
Adding Users: Typeahead search does not populate "Create New User:" fields from LDAP (AD)

### ISSUE THREAD

**Issue by stf675** [osTicket/osTicket-plugins#29]
OSt 1.9.0
PHP 5.5.11
MySQL 5.6.16
(Windows)
LDAPAuth plugin 0.5

When I start using version 1.8.2 I noticed that population of "Create New User" fields only work for few of my AD users)

What was happening, you start typing user name or email in "Search by email, phone or name" field, then click on search results and nothing happened.  Once in a while user information get populated, but most of the time it does not.

I believe I saw mentioning that this will be fixed with 1.9, but when i tried to create a new user, after upgrade the same buggy behavior has occurred.

My search base set to the top of my domain.  The only difference I noticed, is that users that are not "clickable" are under the same parent OU, while users that work are in the other OU.

Saying that, if I add a staff member, - ldap lookup always work regardles what OU that user is in.

he only warning I see in PHP error log is

PHP Warning:  array_key_exists(): The first argument should be either a string or an integer in phar://C:/xampp/htdocs/helpdesk/include/plugins/auth-ldap.phar/include/Net/LDAP2/Entry.php on line 466

**Comment by greezybacon** [comment]
Hmmm. What is the request sent to the backend after you "click" on one of the user?

**Comment by stf675** [comment]
Is there a log that can be enabled/checked?
Or change error logging level?

**Comment by stf675** [comment]
from access.log (replaced real info with XXXX).  What stand out - is 404 status for the 1st line
Pulled info (replaced with XXXX) was correct.

[21/May/2014:12:17:13 -0500] "GET /helpdesk/scp/ajax.php/users/select/auth%3Aldap%3ACN%3DXXXXXX%5C%2C%20XXXXXX%2COU%3DXXXXXXXXX%2COU%3DXXXXXXXX%2COU%3DXXXXXX%20XXXXX%20XXXXX%2COU%3DXXXXXXX%20XXXXXX%2CDC%3DXXXXXXX%2CDC%3DXXXXX HTTP/1.1" 404 1225 "http://XXXXXXXX/helpdesk/scp/users.php" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:29.0) Gecko/20100101 Firefox/29.0"

[21/May/2014:12:17:23 -0500] "GET /helpdesk/scp/ajax.php/users?q=<my search string> HTTP/1.1" 200 247 "http://XXXXXXXX/helpdesk/scp/users.php" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:29.0) Gecko/20100101 Firefox/29.0"

**Comment by stf675** [comment]
still exist in 1.9.1

**Comment by stf675** [comment]
are there any built in restrictions for the length of the user's FQDN?  it may be a reason that some account work and some are not.
with special url encoding - my path could be 140-180 characters long

**Comment by stf675** [comment]
...any update or pointer where to look?  This issue will not help when I do a presentation to upper management  :)

**Comment by stf675** [comment]
Update:

same user account  (that generate 404 error in access.log when you try to add that account via "users" ) can be added without problems as staff member.  After that it can be added as a user without problem.

Also if user send an email that creates a ticket, then that user is also added properly.

**Comment by greezybacon** [comment]
> Also if user send an email that creates a ticket, then that user is also added properly.

That makes sense. We don't yet do LDAP lookup for email ticket + user creation

**Comment by stf675** [comment]
so looks like it has something to do with "fgdn path to user" in my OU structure being too long.
as it works properly for our service and non user accounts that reside on much shorter path

As I stated - it able to see the user, but not populate details.

**Comment by greezybacon** [comment]
I've previously heard that there is something up with the OU's. As far as I know, though, there shouldn't be any truncation.

**Comment by stf675** [comment]
to simplify the example:
we have deeply nested OU for regular users
we have non nested OUs for service accounts and non user accounts

when user (being looked up) is in nested OU, apache access.log generates: 404 error:

GET /helpdesk/scp/ajax.php/users/select/auth%3Aldap%3ACN%3DXXXXXX%5C%2C%20XXXXXX%2COU%3DXXXXXXXXX%2COU%3DXXXXXXXX%2COU%3DXXXXXX%20XXXXX%20XXXXX%2COU%3DXXXXXXX%20XXXXXX%2CDC%3DXXXXXXX%2CDC%3DXXXXX HTTP/1.1" 404 1225 "http://XXXXXXXX/helpdesk/scp/users.php"   (number of X do not represent number of characters, but full length starting with CN%3D is about 170 characters long)

when user is in OU that is not nested, access.log records sucessful completion - 200:

GET /helpdesk/scp/ajax.php/users?q= HTTP/1.1" 200 247 "http://XXXXXXXX/helpdesk/scp/users.php"

**Comment by stf675** [comment]
Did some debugging using chrome:

results:

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<title>Object not found!</title>
<link rev="made" href="mailto:postmaster@localhost" />

<style type="text/css"><![CDATA[/*></style>

</head>
<body>

<h1>Object not found!</h1>

<p>
    The requested URL was not found on this server.
    The link on the
    <a href="http://REMOVED BY ME/helpdesk/scp/users.php">referring
    page</a> seems to be wrong or outdated. Please inform the author of
    <a href="http://REMOVED BY ME/helpdesk/scp/users.php">that page</a>
    about the error.
</p>

<p>
If you think this is a server error, please contact
the <a href="mailto:postmaster@localhost">webmaster</a>.
</p>

<h2>Error 404</h2>

<address>
  <a href="/">MYHOSTNAME</a><br />
  <span>Apache/2.4.9 (Win32) OpenSSL/1.0.1g PHP/5.5.11</span>
</address>
</body>
</html>

headers: (each character from actual query replaced with #"

Remote Address:xx.x.x.xxx:80
Request URL:http://REMOVED BY ME/helpdesk/scp/ajax.php/users/select/auth%3Aldap%3ACN%3#####%5C%2C%20####%2COU%3D###%20#######%2COU%3D##%2COU%3D#######%2COU%3D#####%20####%20##%2COU%3D#########%20#####%2CDC%3D#####%2CDC%3Dcom
Request Method:GET
Status Code:404 Not Found
Request Headersview parsed
GET /helpdesk/scp/ajax.php/users/select/auth%3Aldap%3ACN%3#####%5C%2C%20####%2COU%3D###%20#######%2COU%3D##%2COU%3D#######%2COU%3D#####%20####%20##%2COU%3D#########%20#####%2CDC%3D#####%2CDC%3Dcom HTTP/1.1
Host: MYHOSTNAME
Connection: keep-alive
Accept: text/html, _/_; q=0.01
X-Requested-With: XMLHttpRequest
User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.153 Safari/537.36
Referer: http://MYHOSTNAME/helpdesk/scp/users.php
Accept-Encoding: gzip,deflate,sdch
Accept-Language: en-US,en;q=0.8
Cookie: OSTSESSID=f6ft2q785kmt6r12p5jacui1e7
Response Headersview parsed
HTTP/1.1 404 Not Found
Date: Thu, 19 Jun 2014 01:31:39 GMT
Server: Apache/2.4.9 (Win32) OpenSSL/1.0.1g PHP/5.5.11
Vary: accept-language,accept-charset
Accept-Ranges: bytes
Keep-Alive: timeout=5, max=99
Connection: Keep-Alive
Transfer-Encoding: chunked
Content-Type: text/html; charset=utf-8
Content-Language: en

**Comment by stf675** [comment]
issue still exist in 1.9.4

**Comment by greezybacon** [comment]
> Vary: accept-language,accept-charset

Interesting. Our code does not add the `Vary` header to the response. Do you have an Apache rule to do so? I'm wondering if Apache is not actually serving this request to osTicket.

**Comment by stf675** [comment]
That what shows up in Firefox "web developer" section.

I updated plugin and osTicket version the the latest - problem still exist.

some OUs return data, some fail.

Failing request URL (actual name replaced with \* - same number of characters)

http://**_**-__-__/_******_/scp/ajax.php/users/select/auth%3Aldap%3ACN%3D**_*****_%5C%2C%20**_*****_%2COU%3D**_**_%20**__%2COU%3D__%2COU%3D_*****_%2COU%3D**_**%20****%20**%2COU%3D********_%20**_**%2CDC%3D*****%2CDC%3Dcom

Request method GET
Status code: 404 Not Found

**Comment by stf675** [comment]
This may help to fix it (hopefully)

I tried an account from same OU, that does not contain comma in display name and it is work,  Looks like comma in display name (CN=)  breaks it. 

similar issue described at http://osticket.com/forum/discussion/85048/auth-ldap-issue-with-cn-containing-comma/p1

Update:

comma does not matter when adding Agents using LDAP, only when adding users.

**Comment by sjlevy** [comment]
I am the author of the forum post: http://osticket.com/forum/discussion/85048/auth-ldap-issue-with-cn-containing-comma

I too would like to see this bug examined, unfortunately all of our users have commas in their CN's making this a bother

Could there be a double escaping issue?  Perhaps the comma being escaped once in the auth-ldap/osticket ajax/user handling, and then again by PEAR's Net_LDAP2, causing lookup to fail?

**Comment by stf675** [comment]
would be interesting to examine what the difference when you create Agent, vs User.  Because Agent creation works when name contained comma. (but it is a bit above my level)

**Comment by sjlevy** [comment]
I found that $lookup_dn on line 291 of auth-ldap/authentication.php has a forward slash instead of a backslash escape-- causing the issue (AD query returns no results)

I haven't been able to locate the source of the problem yet, but a simple band aid is to change line 291

From: `$r = $c->search($lookup_dn, '(objectClass=*)', $opts);`
To: `$r = $c->search(str_replace('/','\\',$lookup_dn), '(objectClass=*)', $opts);`

To edit the auth-ldap plugin code you will need to first:
- Make sure you have your LDAP credentials on hand and a local admin account/password for accessing OST (non-remote)
- Disable and Delete your LDAP plugin under Admin Panel/Manage/Plugins
- Download the auth-ldap folder from this core-plugins repository
- Move it into osTicket's plugin folder
- Delete the old auth-ldap.phar
- Download the Net_LDAP2 dependency https://pear.php.net/package/Net_LDAP2/redirected and move the Net folder into osTicket's include/plugins/auth-ldap/include/
- Log in to osTicket using the local admin account
- Add/Configure/Enable the auth-ldap plugin (now using the folder instead of the phar)
- Make the line change in auth-ldap/authentication.php
- Verify remote login works and new ticket lookups with comma in CN

@greezybacon can likely come up with a better solution in a future update

**Comment by greezybacon** [comment]
Can anyone say if this patch fixes the issue?

``` patch
diff --git a/auth-ldap/authentication.php b/auth-ldap/authentication.php
index 34d78ea..2edd3f1 100644
--- a/auth-ldap/authentication.php
+++ b/auth-ldap/authentication.php
@@ -288,6 +288,7 @@ class LDAPAuthentication {
                 $schema['username'],
             )))
         );
+        $lookup_dn = Net_LDAP2_Util::canonical_dn($lookup_dn);
         $r = $c->search($lookup_dn, '(objectClass=*)', $opts);
         if (PEAR::isError($r) || !$r->count())
             return null;
```

**Comment by ShaneOss** [comment]
As stf675 says..
When adding an Agent using type-ahead search and clicking on a username from LDAP it auto populates First Name, Last Name and Email.

Couldn't we just use that function for auto populating client data too?

Also greezybacon patch didn't fix the issue. :)

**Comment by ShaneOss** [comment]
OK after spending about a few days on this now I found the problem and a fix.
sjlevy is right but instead of modifying the LDAP authentication plugin. Modify the file include\ajax.users.php

Insert the following line in the function addRemoteUser:
INSERT CODE: $id = str_replace('/','\\',$id);
Above Line: if (!$thisstaff)

So your removing the forward slash and replacing it with a backslash.

Cheers,
Shane

**Comment by stf675** [comment]
Hi ShaneOss,

adding `$id = str_replace('/','\\',$id);`
seems to be working.  

Thank you.

**Comment by DarthSlider** [comment]
@ShaneOss 
Hi! Looks like it doesn't work because "\" symbol counts like a start of the comment and should be shielded somehow. I'm not good at php so don't know solution, but then i add this string to ajax.users.php file in notepad++ everything beneath this string became grayed like a comment.

**Comment by DarthSlider** [comment]
@ShaneOss Ok, i found that i should use` $id = str_replace('/','\\',$id); `string, but typeahead search still doesn't work. Maybe it's because i use php7.1?

**Comment by DarthSlider** [comment]
@ShaneOss 
Found this solution: https://github.com/osTicket/osTicket/pull/3456
It helps.

**Comment by ShaneOss** [comment]
@DarthSlider Thanks. I fixed my issue with https://github.com/osTicket/osTicket-plugins/issues/29#issuecomment-218035136
Cheers

**Comment by ShaneOss** [comment]
@DarthSlider You'd need to provide some more details about your environment to be able to assist, if its still not working for you. Could be configuration. We're not using PHP 7.1. Can you logon using LDAP account?

**Comment by sevmonster** [comment]
This is still an issue in latest develop branch. Despite knowing the reason (comma in CN) this has not been fixed in 6 years...

**Comment by sevmonster** [comment]
https://github.com/osTicket/osTicket-plugins/issues/29#issuecomment-126546298 did not work, but https://github.com/osTicket/osTicket-plugins/issues/29#issuecomment-114178879 did.
I imagine this might have something to do with URL normalizing somewhere down the pipe. I can confirm that my instance of Chrome is sending %5C in the request URL, so it is getting malformed somewhere inside osTicket.

Edit: Looked into it a tad bit more, and like @ShaneOss discovered, the issue is not in the auth plugin. The malformed DN comes in directly to `StaffLDAPAuthentication.lookup`. I ended up inserting this line before the `_ldap->lookup` call in the above function:
`$dn = str_replace('/', '\\', $dn);`

**Comment by eggi36** [comment]
Even to this day the issue persists with the following versions:

osTicket-Version | v1.15.4 (6bd7884)
Server-Software | Apache/2.4.51 (Win64) OpenSSL/1.1.1l PHP/7.4.26
MySQL-Version | 10.4.22
PHP-Version | 7.4.26
LDAP Plugin | 0.6.2

I've tried all the workaround listed in this issue but none of them works. As soon as a comma is in the CN the auto population of the "create new user"dialog does not work.

**Comment by sevmonster** [comment]
> I've tried all the workaround listed in this issue but none of them works. As soon as a comma is in the CN the auto population of the "create new user"dialog does not work.

Did you try my change?
> Looked into it a tad bit more, and like @ShaneOss discovered, the issue is not in the auth plugin. The malformed DN comes in directly to `StaffLDAPAuthentication.lookup`. I ended up inserting this line before the `_ldap->lookup` call in the above function: `$dn = str_replace('/', '\\', $dn);`

**Comment by eggi36** [comment]
Yes. I've added the line in authentication.php here:

```
    function lookup($dn) {
	$dn = str_replace('/', '\\', $dn);
        $hit =  $this->_ldap->lookup($dn);
```

Sadly it still not works.

**Comment by sevmonster** [comment]
@eggi36 It will be a while but once I get back into work I can provide a patch for you to try.

**Comment by eggi36** [comment]
@sevmonster That would be awesome! Let me know if you need any specific logs etc.

**Comment by sevmonster** [comment]
@eggi36 Sorry it took so long, so I don't know if you're still encountering this issue, but I just checked our patches and they are essentially identical to what is here.
So if it isn't working for you, it must be some other issue.
Note that we use MS AD for our LDAP, if you are using another server you may be encountering another issue or incompatibility.

**Comment by eggi36** [comment]
Thank you for asking back! To be honest: It is working now. It suddenly started to work a few days after I posted my issue here, I don't know which changes I made were responsible for it in the end.
